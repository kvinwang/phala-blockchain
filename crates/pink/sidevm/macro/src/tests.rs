#[test]
fn test_ocall() {
    let stream = crate::macro_ocall::patch(syn::parse_quote! {
        pub trait Ocall {
            #[ocall(id = 101)]
            fn call_slow(a: i32, b: i32) -> i32;

            #[ocall(id = 103, fast_input)]
            fn call_fi(a: i32, b: i32) -> i32;

            #[ocall(id = 104, fast_return)]
            fn call_fo(a: i32, b: i32) -> i32;

            #[ocall(id = 102, fast_input, fast_return)]
            fn poll_fi_fo(a: i32, b: i32) -> i32;
        }
    });
    insta::assert_display_snapshot!(rustfmt_snippet::rustfmt_token_stream(&stream).unwrap())
}

#[test]
fn test_main() {
    let stream = crate::macro_main::patch(syn::parse_quote! {
        async fn the_main() {
            sleep(1).await
        }
    });
    insta::assert_display_snapshot!(rustfmt_snippet::rustfmt_token_stream(&stream).unwrap())
}
