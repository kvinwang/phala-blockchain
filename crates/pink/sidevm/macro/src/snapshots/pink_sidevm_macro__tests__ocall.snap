---
source: crates/pink/sidevm/macro/src/tests.rs
assertion_line: 18
expression: "rustfmt_snippet::rustfmt_token_stream(&stream).unwrap()"
---
pub trait Ocall {
    fn call_slow(&mut self, a: i32, b: i32) -> i32;
    fn call_fi(&mut self, a: i32, b: i32) -> i32;
    fn call_fo(&mut self, a: i32, b: i32) -> i32;
    fn poll_fi_fo(&mut self, a: i32, b: i32) -> i32;
}
pub mod ocall_guest {
    use super::*;
    pub fn call_slow(a: i32, b: i32) -> i32 {
        unsafe {
            let inputs = (a, b);
            let mut input_buf = empty_buffer();
            Encode::encode_to(&inputs, &mut input_buf);
            let len = input_buf.len() as IntPtr;
            let ret = sidevm_ocall(current_task(), 101, input_buf.as_ptr() as IntPtr, len, 0, 0);
            let len = <Result<i32> as RetDecode>::decode_ret(ret)?;
            if len < 0 {
                panic!("ocall returned an error");
            }
            let mut buf = alloc_buffer(len as _);
            let ret = sidevm_ocall_fast_return(
                current_task(),
                0,
                buf.as_mut_ptr() as IntPtr,
                len as IntPtr,
                0,
                0,
            );
            let ret = <Result<i32> as RetDecode>::decode_ret(ret)?;
            if ret != len {
                panic!("ocall get return length mismatch");
            }
            Ok(Decode::decode(&mut buf.as_ref()).expect("Failed to decode ocall return value"))
        }
    }
    pub fn call_fi(a: i32, b: i32) -> i32 {
        unsafe {
            let stack = StackedArgs::empty();
            let stack = stack.push_arg(a);
            let stack = stack.push_arg(b);
            let args = stack.dump();
            let ret = sidevm_ocall(current_task(), 103, args[0], args[1], args[2], args[3]);
            let len = <Result<i32> as RetDecode>::decode_ret(ret)?;
            if len < 0 {
                panic!("ocall returned an error");
            }
            let mut buf = alloc_buffer(len as _);
            let ret = sidevm_ocall_fast_return(
                current_task(),
                0,
                buf.as_mut_ptr() as IntPtr,
                len as IntPtr,
                0,
                0,
            );
            let ret = <Result<i32> as RetDecode>::decode_ret(ret)?;
            if ret != len {
                panic!("ocall get return length mismatch");
            }
            Ok(Decode::decode(&mut buf.as_ref()).expect("Failed to decode ocall return value"))
        }
    }
    pub fn call_fo(a: i32, b: i32) -> i32 {
        unsafe {
            let inputs = (a, b);
            let mut input_buf = empty_buffer();
            Encode::encode_to(&inputs, &mut input_buf);
            let len = input_buf.len() as IntPtr;
            let ret = sidevm_ocall_fast_return(
                current_task(),
                104,
                input_buf.as_ptr() as IntPtr,
                len,
                0,
                0,
            );
            <Result<i32> as RetDecode>::decode_ret(ret).map(I32Convertible::from_i32)
        }
    }
    pub fn poll_fi_fo(a: i32, b: i32) -> i32 {
        unsafe {
            let stack = StackedArgs::empty();
            let stack = stack.push_arg(a);
            let stack = stack.push_arg(b);
            let args = stack.dump();
            let ret =
                sidevm_ocall_fast_return(current_task(), 102, args[0], args[1], args[2], args[3]);
            <Result<i32> as RetDecode>::decode_ret(ret).map(I32Convertible::from_i32)
        }
    }
}
pub fn dispatch_call_fast_return<Env: Ocall + OcallEnv, Vm: VmMemory>(
    env: &mut Env,
    vm: &Vm,
    id: i32,
    p0: IntPtr,
    p1: IntPtr,
    p2: IntPtr,
    p3: IntPtr,
) -> Result<i32> {
    match id {
        0 => {
            let buffer = env.take_return().ok_or(OcallError::NoReturnValue)?;
            let len = p1 as usize;
            if buffer.len() != len {
                return Err(OcallError::InvalidParameter);
            }
            vm.copy_to_vm(&buffer, p0)?;
            Ok(len as i32)
        }
        104 => {
            let (a, b) = {
                let mut buf = vm.slice_from_vm(p0, p1)?;
                Decode::decode(&mut buf).or(Err(OcallError::InvalidParameter))?
            };
            env.call_fo(a, b).map(I32Convertible::to_i32)
        }
        102 => {
            let stack = StackedArgs::load(&[p0, p1, p2, p3]).ok_or(OcallError::InvalidParameter)?;
            let (b, stack) = stack.pop_arg(vm)?;
            let (a, stack) = stack.pop_arg(vm)?;
            let _: StackedArgs<()> = stack;
            env.poll_fi_fo(a, b).map(I32Convertible::to_i32)
        }
        _ => Err(OcallError::UnknownCallNumber),
    }
}
pub fn dispatch_call<Env: Ocall + OcallEnv, Vm: VmMemory>(
    env: &mut Env,
    vm: &Vm,
    id: i32,
    p0: IntPtr,
    p1: IntPtr,
    p2: IntPtr,
    p3: IntPtr,
) -> Result<i32> {
    Ok(match id {
        0 => {
            let ret = {
                let buffer = env.take_return().ok_or(OcallError::NoReturnValue)?;
                let len = p1 as usize;
                if buffer.len() != len {
                    return Err(OcallError::InvalidParameter);
                }
                vm.copy_to_vm(&buffer, p0)?;
                Ok(len as i32)
            };
            env.put_return(ret?.encode()) as _
        }
        101 => {
            let (a, b) = {
                let mut buf = vm.slice_from_vm(p0, p1)?;
                Decode::decode(&mut buf).or(Err(OcallError::InvalidParameter))?
            };
            let ret = env.call_slow(a, b);
            env.put_return(ret?.encode()) as _
        }
        103 => {
            let stack = StackedArgs::load(&[p0, p1, p2, p3]).ok_or(OcallError::InvalidParameter)?;
            let (b, stack) = stack.pop_arg(vm)?;
            let (a, stack) = stack.pop_arg(vm)?;
            let _: StackedArgs<()> = stack;
            let ret = env.call_fi(a, b);
            env.put_return(ret?.encode()) as _
        }
        104 => {
            let (a, b) = {
                let mut buf = vm.slice_from_vm(p0, p1)?;
                Decode::decode(&mut buf).or(Err(OcallError::InvalidParameter))?
            };
            let ret = env.call_fo(a, b);
            env.put_return(ret?.encode()) as _
        }
        102 => {
            let stack = StackedArgs::load(&[p0, p1, p2, p3]).ok_or(OcallError::InvalidParameter)?;
            let (b, stack) = stack.pop_arg(vm)?;
            let (a, stack) = stack.pop_arg(vm)?;
            let _: StackedArgs<()> = stack;
            let ret = env.poll_fi_fo(a, b);
            env.put_return(ret?.encode()) as _
        }
        _ => return Err(OcallError::UnknownCallNumber),
    })
}
pub fn ocall_id2name(id: i32) -> &'static str {
    match id {
        0 => "get_return",
        101i32 => "call_slow",
        103i32 => "call_fi",
        104i32 => "call_fo",
        102i32 => "poll_fi_fo",
        _ => "unknown",
    }
}

